velox/functions/prestosql/tests/GreatestLeastTest.cpp

```diff
@@ -83,29 +84,52 @@ TEST_F(GreatestLeastTest, leastReal) {
       {0, -100, -1.1});
 }
 
-TEST_F(GreatestLeastTest, nanInput) {
-  // Presto rejects NaN inputs of type DOUBLE, but allows NaN inputs of type
-  // REAL.
-  std::vector<double> input{0, 1.1, std::nan("1")};
-  VELOX_ASSERT_THROW(
-      runTest<double>("least(c0)", {{0.0 / 0.0}}, {0}),
-      "Invalid argument to least(): NaN");
-  runTest<double>("try(least(c0, 1.0))", {input}, {0, 1.0, std::nullopt});
-
-  VELOX_ASSERT_THROW(
-      runTest<double>("greatest(c0)", {1, {0.0 / 0.0}}, {1, 0}),
-      "Invalid argument to greatest(): NaN");
-  runTest<double>("try(greatest(c0, 1.0))", {input}, {1.0, 1.1, std::nullopt});
-
-  auto result = evaluateOnce<bool, float, float>(
-      "is_nan(least(c0))", std::nanf("1"), 1.2);
-  ASSERT_TRUE(result.has_value());
-  ASSERT_TRUE(result.value());
-
-  result = evaluateOnce<bool, float, float>(
-      "is_nan(greatest(c0))", std::nanf("1"), 1.2);
-  ASSERT_TRUE(result.has_value());
-  ASSERT_TRUE(result.value());
+TEST_F(GreatestLeastTest, greatestNanInput) {
+  auto greatestFloatTestThreeArgs = [&](float a, float b, float c) {
+    return evaluateOnce<float, float, float, float>(
+               "greatest(c0, c1, c2)", {a}, {b}, {c})
+        .value();
+  };
+
+  auto greatestDoubleTestThreeArgs = [&](double a, double b, double c) {
+    return evaluateOnce<double, double, double, double>(
+               "greatest(c0, c1, c2)", {a}, {b}, {c})
+        .value();
+  };
+
+  EXPECT_TRUE(std::isnan(greatestFloatTestThreeArgs(1.0, std::nanf("1"), 2.0)));
+  EXPECT_TRUE(std::isnan(greatestFloatTestThreeArgs(
+      std::nanf("1"), 1.0, std::numeric_limits<float>::infinity())));
+
+  EXPECT_TRUE(std::isnan(greatestDoubleTestThreeArgs(1.0, std::nan("1"), 2.0)));
+  EXPECT_TRUE(std::isnan(greatestDoubleTestThreeArgs(
+      std::nan("1"), 1.0, std::numeric_limits<double>::infinity())));
+}
+
+TEST_F(GreatestLeastTest, leastNanInput) {
+  auto leastFloatTestThreeArgs = [&](float a, float b, float c) {
+    return evaluateOnce<float, float, float, float>(
+               "least(c0, c1, c2)", {a}, {b}, {c})
+        .value();
+  };
+
+  auto leastDoubleTestThreeArgs = [&](double a, double b, double c) {
+    return evaluateOnce<double, double, double, double>(
+               "least(c0, c1, c2)", {a}, {b}, {c})
+        .value();
+  };
+
+  EXPECT_EQ(leastFloatTestThreeArgs(1.0, std::nanf("1"), 0.5), 0.5);
+  EXPECT_EQ(
+      leastFloatTestThreeArgs(
+          std::nanf("1"), 1.0, -std::numeric_limits<float>::infinity()),
```

Comment:
nit: use kNegativeInf32 and kNegativeInf64 (or similar) constants for readability